packages/rules-unit-testing/src/impl/discovery.ts

```diff
@@ -88,25 +88,23 @@ export function getEmulatorHostAndPort(
   discovered?: DiscoveredEmulators
 ) {
   if (conf && ('host' in conf || 'port' in conf)) {
-    const { host, port } = conf;
-    if (host || port) {
-      if (!host || !port) {
-        throw new Error(
-          `Invalid configuration ${emulator}.host=${host} and ${emulator}.port=${port}. ` +
-            'If either parameter is supplied, both must be defined.'
-        );
-      }
-      if (discovered && !discovered[emulator]) {
-        console.warn(
-          `Warning: config for the ${emulator} emulator is specified, but the Emulator hub ` +
-            'reports it as not running. This may lead to errors such as connection refused.'
-        );
-      }
-      return {
-        host: fixHostname(conf.host, discovered?.hub?.host),
-        port: conf.port
-      };
+    const { host, port } = conf as any;
+    if (!host || !port) {
```

Comment:
nit: If the goal of this PR is to upgrade TypeScript and fix compile errors, then the only diff required here is `as any` above. I'm not opposed to your other changes but I think they belong to a different PR. Please revert this line and below to keep the PR clean.

Note: Removing the outer `if (host || port) {` is technically a small behavior change that only matters for `{port: undefined}` or `{port: null}` (or ditto for host), but it has the potential to break some customers.

Reply:
I should've reverted the other changes in the first place. I've now reverted most of the other changes. I made one small change [here](https://github.com/firebase/firebase-js-sdk/pull/8561/files#diff-3faece18a369c358e4c5072f7c97ba4713bb8a0af322eb698f00cef7b3cd9329L106-L107), because TS now deduces that `conf` is of type 
```ts
conf: {
    rules?: string | undefined;
} & Record<"port", unknown>
```
so `conf.host` does not exist. It's not clear to me why this is the case. Let me know if you have any objections to just using `host` and `port` instead of `conf.host` and `conf.port` to get around this issue.

Reply:
No objections at all. Thanks for handling this with care.